// @atrilabs/design-system/lib/consts/colors.js
// gray
export var white = "#ffffff";
export var gray50 = "#F9FAFB";
export var gray100 = "#f3f4f6";
export var gray200 = "#e5e7eb";
export var gray300 = "#d1d5db";
export var gray400 = "#9ca3af";
export var gray500 = "#6b7280";
export var gray600 = "#4b5563";
export var gray700 = "#374151";
export var gray800 = "#1f2937";
export var gray800HalfOpacity = "rgba(31, 41, 55, 0.5)";
export var gray900 = "#111827";
// amber
export var amber200 = "#fde68a";
export var amber300 = "#fcd34d";
// sky
export var sky200 = "#bae6fd";
export var sky400 = "#38BDF8";
export var sky500 = "#0ea5e9";
export var sky800 = "#075985";
// teal
export var teal400 = "#2DD4BF";
// inputLiner-Grad
export var inputgrad = "linear-gradient(0deg, rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2)), #1f2937";
// agastyaLine
export var agastyaLine = "rgba(31, 41, 55, 0.5)";
// orange
export var orange200 = "#FED7AA";
export var orange300 = "#FDBA74";
export var orange400 = "#FB923C";
export var orange600 = "#EA580C";
// pink
export var pink600 = "#DB2777";
export var pink700 = "#BE185D";
// blue
export var blue600 = "#2563EB";
