// @atrilabs/design-system/lib/consts/typography.js
export var Inter = '"Inter", sans-serif';
export var Logo = '"Lobster", cursive';
// typography
export var h1Heading = {
    fontFamily: Inter,
    fontStyle: "normal",
    fontWeight: 600,
    fontSize: "1.125rem",
    lineHeight: "22px",
    color: "#000000",
};
export var h3Heading = {
    fontFamily: Inter,
    fontSize: "0.875rem",
    fontStyle: "normal",
    fontWeight: 600,
    lineHeight: "17px",
    letterSpacing: "0em",
};
export var h4Heading = {
    fontFamily: Inter,
    fontStyle: "normal",
    fontWeight: 500,
    fontSize: "11px",
    lineHeight: "13px",
    color: "#000000",
};
export var h5Heading = {
    fontFamily: Inter,
    fontStyle: "normal",
    fontWeight: "bold",
    fontSize: "11px",
    lineHeight: "13px",
    color: "#000000",
};
export var smallText = {
    fontFamily: Inter,
    fontStyle: "normal",
    fontWeight: "normal",
    fontSize: "11px",
    lineHeight: "13px",
    color: "#000000",
};
export var body = {
    fontFamily: Inter,
    fontStyle: "normal",
    fontWeight: "normal",
    fontSize: "0.75rem",
    lineHeight: "15px",
};
