// @atrilabs/scripts/build/shared/build-packages.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRuntime = exports.buildLayer = void 0;
var child_process_1 = require("child_process");
var command_exists_1 = __importDefault(require("command-exists"));
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
/**
 * Strategy -
 * Find tsc in node_modules/typescript/bin/tsc.
 * If not found, then find if the command exists globally.
 */
function findLocal() {
    var found = true;
    try {
        // if throws error, then typscript isn't installed locally
        require.resolve("typescript/bin/tsc");
    }
    catch (err) {
        found = false;
    }
    return found;
}
function findGlobal() {
    return command_exists_1.default.sync("tsc");
}
/**
 *
 * @param cmd either output of require.resolve(typescript/bin/tsc) or tsc
 */
function runTsc(cwd, cmd) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (res, rej) {
                    (0, child_process_1.exec)(cmd, { cwd: cwd }, function (err, stdout, stderr) {
                        if (err) {
                            console.log("err:\n".concat(err));
                        }
                        if (stderr) {
                            console.log("stderr:\n".concat(stderr));
                        }
                        if (stdout) {
                            console.log("stdout:\n".concat(stdout));
                        }
                        if (err === null || err.code === 0) {
                            res();
                        }
                        else {
                            rej("tsc exited with with some error\n ".concat(err));
                        }
                    });
                })];
        });
    });
}
function readTsConfig(cwd) {
    var rootDir = cwd;
    var tsconfigPath = path_1.default.resolve(rootDir, "tsconfig.json");
    if (fs_1.default.existsSync(tsconfigPath)) {
        var json = JSON.parse(fs_1.default.readFileSync(tsconfigPath).toString());
        return {
            rootDir: path_1.default.resolve(rootDir, json["compilerOptions"]["rootDir"]),
            outDir: path_1.default.resolve(rootDir, json["compilerOptions"]["outDir"]),
        };
    }
    return;
}
function getFiles(dir) {
    return __asyncGenerator(this, arguments, function getFiles_1() {
        var dirents, _i, dirents_1, dirent, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, __await(fs_1.default.promises.readdir(dir, { withFileTypes: true }))];
                case 1:
                    dirents = _a.sent();
                    _i = 0, dirents_1 = dirents;
                    _a.label = 2;
                case 2:
                    if (!(_i < dirents_1.length)) return [3 /*break*/, 9];
                    dirent = dirents_1[_i];
                    res = path_1.default.resolve(dir, dirent.name);
                    if (!dirent.isDirectory()) return [3 /*break*/, 5];
                    return [5 /*yield**/, __values(__asyncDelegator(__asyncValues(getFiles(res))))];
                case 3: return [4 /*yield*/, __await.apply(void 0, [_a.sent()])];
                case 4:
                    _a.sent();
                    return [3 /*break*/, 8];
                case 5: return [4 /*yield*/, __await(res)];
                case 6: return [4 /*yield*/, _a.sent()];
                case 7:
                    _a.sent();
                    _a.label = 8;
                case 8:
                    _i++;
                    return [3 /*break*/, 2];
                case 9: return [2 /*return*/];
            }
        });
    });
}
function _copyAssetsFromSrc(rootDir, outDir) {
    var e_1, _a;
    return __awaiter(this, void 0, void 0, function () {
        var assetRegex, toBeCopied, _b, _c, f, e_1_1, copyPromises;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    assetRegex = /(\.avif$)|(\.bmp$)|(\.gif$)|(\.jpe?g$)|(\.png$)|(\.svg$)|(\.css$)|(\.(scss|sass)$)/g;
                    toBeCopied = [];
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 6, 7, 12]);
                    _b = __asyncValues(getFiles(rootDir));
                    _d.label = 2;
                case 2: return [4 /*yield*/, _b.next()];
                case 3:
                    if (!(_c = _d.sent(), !_c.done)) return [3 /*break*/, 5];
                    f = _c.value;
                    if (f.match(assetRegex)) {
                        toBeCopied.push(f);
                    }
                    _d.label = 4;
                case 4: return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 12];
                case 6:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 12];
                case 7:
                    _d.trys.push([7, , 10, 11]);
                    if (!(_c && !_c.done && (_a = _b.return))) return [3 /*break*/, 9];
                    return [4 /*yield*/, _a.call(_b)];
                case 8:
                    _d.sent();
                    _d.label = 9;
                case 9: return [3 /*break*/, 11];
                case 10:
                    if (e_1) throw e_1.error;
                    return [7 /*endfinally*/];
                case 11: return [7 /*endfinally*/];
                case 12:
                    copyPromises = [];
                    toBeCopied.forEach(function (f) {
                        var promise = fs_1.default.promises.readFile(f).then(function (buf) {
                            var dest = path_1.default.resolve(outDir, path_1.default.relative(rootDir, f));
                            if (!fs_1.default.existsSync(path_1.default.dirname(dest))) {
                                fs_1.default.mkdirSync(path_1.default.dirname(dest), { recursive: true });
                            }
                            return fs_1.default.promises.writeFile(dest, buf);
                        });
                        copyPromises.push(promise);
                    });
                    return [4 /*yield*/, Promise.all(copyPromises)];
                case 13:
                    _d.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function copyAssetsFromSrc(cwd) {
    var tsconfig = readTsConfig(cwd);
    // by default we will put assets in outDir
    var outDir = path_1.default.resolve(cwd, "lib");
    var rootDir = path_1.default.resolve(cwd, "src");
    if (tsconfig) {
        rootDir = tsconfig.rootDir;
        outDir = tsconfig.outDir;
    }
    _copyAssetsFromSrc(rootDir, outDir);
}
function buildLayer(cwd) {
    if (findLocal()) {
        var cmd = "node ".concat(require.resolve("typescript/bin/tsc"));
        return runTsc(cwd, cmd)
            .then(function () {
            copyAssetsFromSrc(cwd);
        })
            .catch(function (err) { return console.log(err); });
    }
    else if (findGlobal()) {
        var cmd = "tsc";
        return runTsc(cwd, cmd)
            .then(function () {
            copyAssetsFromSrc(cwd);
        })
            .catch(function (err) { return console.log(err); });
    }
    else {
        throw Error("'typescript' package is missing. Please install typescript package.");
    }
}
exports.buildLayer = buildLayer;
function buildRuntime(cwd) {
    if (findLocal()) {
        var cmd = "node ".concat(require.resolve("typescript/bin/tsc"));
        return runTsc(cwd, cmd)
            .then(function () {
            copyAssetsFromSrc(cwd);
        })
            .catch(function (err) { return console.log(err); });
    }
    else if (findGlobal()) {
        var cmd = "tsc";
        return runTsc(cwd, cmd)
            .then(function () {
            copyAssetsFromSrc(cwd);
        })
            .catch(function (err) { return console.log(err); });
    }
    else {
        throw Error("'typescript' package is missing. Please install typescript package.");
    }
}
exports.buildRuntime = buildRuntime;
