// @atrilabs/app-scripts/lib/tasks/build-app.js
#!/usr/bin/env node
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var utils_1 = require("../shared/utils");
// @ts-ignore
global.window = undefined;
var mode = (0, utils_1.getMode)();
(0, utils_1.setNodeAndBabelEnv)(mode);
try {
    var buildInfo = JSON.parse(fs_1.default.readFileSync(utils_1.buildInfoFile).toString());
    if (buildInfo) {
        var appEntry = buildInfo["appEntry"];
        var appHtml = buildInfo["appHtml"];
        var appOutput = buildInfo["appOutput"];
        var appSrc = buildInfo["appSrc"];
        var manifestDirs = buildInfo["manifestDirs"];
        if (!(appEntry &&
            typeof appEntry === "string" &&
            appHtml &&
            typeof appHtml === "string" &&
            appOutput &&
            typeof appOutput === "string" &&
            appSrc &&
            typeof appSrc === "string" &&
            manifestDirs &&
            Array.isArray(manifestDirs))) {
            throw Error("Wrong schema of ".concat(utils_1.buildInfoFilename, "."));
        }
        var manifestPkgs = manifestDirs.map(function (manifestDef) {
            if (manifestDef["pkg"]) {
                return path_1.default.dirname(require.resolve("".concat(manifestDef.pkg, "/package.json")));
            }
            else {
                throw Error("Wrong schema of ".concat(utils_1.buildInfoFilename, "."));
            }
        });
        var includes = __spreadArray(__spreadArray([], manifestPkgs, true), [appSrc], false);
        (0, utils_1.buildApp)({
            appEntry: path_1.default.resolve(appEntry),
            appHtml: path_1.default.resolve(appHtml),
            appOutput: path_1.default.resolve(appOutput),
            includes: includes.map(function (inc) { return path_1.default.resolve(inc); }),
            mode: mode,
            addWatchOptions: false,
            assetUrlPrefix: process.env["ASSET_URL_PREFIX"] || buildInfo.assetUrlPrefix,
        });
    }
    else {
        console.log("Missing manifestDirs in ".concat(utils_1.buildInfoFilename));
    }
}
catch (err) {
    console.log("Error\n", err);
}
