// @atrilabs/forest/lib/implementations/lowdb/LowDbForestManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEventManagers = void 0;
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var LowDbEventManager_1 = __importDefault(require("./LowDbEventManager"));
function getForestDirPath(forestDir, forestname) {
    return path_1.default.resolve(forestDir, "forests", forestname);
}
function checkIfDirExists(forestDir, forestname) {
    if (fs_1.default.existsSync(getForestDirPath(forestDir, forestname))) {
        return true;
    }
    return false;
}
// create one directory and event manager for each of forest
function createEventManagers(forestDir, forestsConfig, mode) {
    var forestnames = Object.keys(forestsConfig);
    var eventManagers = {};
    forestnames.forEach(function (forestname) {
        // create directory if does not exist already
        if (!checkIfDirExists(forestDir, forestname)) {
            fs_1.default.mkdirSync(getForestDirPath(forestDir, forestname), {
                recursive: true,
            });
        }
        // create event manager
        var eventManager = (0, LowDbEventManager_1.default)({
            dbDir: getForestDirPath(forestDir, forestname),
            mode: mode,
        });
        eventManagers[forestname] = eventManager;
    });
    return eventManagers;
}
exports.createEventManagers = createEventManagers;
function createLowDbForestManager(options) {
    var eventManagers = createEventManagers(options.forestDir, options.forestsConfig, options.mode);
    function getEventManager(name) {
        if (eventManagers[name] === undefined)
            throw Error("Event Manager for forest type ".concat(name, " doesn't exist. Please check your tool.config.json."));
        return eventManagers[name];
    }
    return { getEventManager: getEventManager };
}
exports.default = createLowDbForestManager;
